extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -228,67 +392,430 @@ class AmpDatePicker extends AMP.BaseElement {
+    this.listen_(documentElement, 'keydown', this.handleKeydown_.bind(this));
+  }
+
+  /**
+   * True if the input is a field of this date picker.
+   * @param {?Element} field
+   * @return {boolean}
+   */
+  isDateField_(field) {
+    return (
+      field === this.dateField_ ||
+      field === this.startDateField_ ||
+      field === this.endDateField_
+    );
+  }
+
+  /**
+   * Handle clicks inside and outside of the date picker to detect when to
+   * open and close the date picker.
+   * @param {!Event} e
+   */
+  handleClick_(e) {
+    // Disregard spacebar and enter "clicks"
+    if (e.screenX == 0 && e.screenY == 0) {
+      return;
+    }
+
+    const target = dev().assertElement(e.target);
+    const clickWasInDatePicker = (
+      this.container_.contains(target) || this.isDateField_(target)
+    );
+
+    if (!clickWasInDatePicker) {
+      this.transitionTo_(DatePickerState.OVERLAY_CLOSED);
+    }
+  }
+
+  /**
+   * Handle focus events in the document to switch between selecting the start
+   * and end dates, and when to open and close the date picker.
+   * @param {!Event} e
+   */
+  handleFocus_(e) {
+    const target = dev().assertElement(e.target);
+    if (this.isDateField_(target)) {
+      if (target == this.startDateField_) {
+        this.updateDateFieldFocus_(this.startDateField_);
+        this.setState_({focusedInput: this.ReactDatesConstants_.START_DATE});
+      } else if (target == this.endDateField_) {
+        this.updateDateFieldFocus_(this.endDateField_);
+        this.setState_({focusedInput: this.ReactDatesConstants_.END_DATE});
+      } else if (target == this.dateField_) {
+        this.updateDateFieldFocus_(this.dateField_);
+      }
+      this.transitionTo_(DatePickerState.OVERLAY_OPEN_INPUT);
+    } else if (this.element.contains(target)) {
+      this.transitionTo_(DatePickerState.OVERLAY_OPEN_PICKER);
+    } else {
+      this.updateDateFieldFocus_(null);
+      this.transitionTo_(DatePickerState.OVERLAY_CLOSED);
+    }
+  }
+
+  /**
+   * For inputs that are valid dates, update the date-picker value.
+   * @param {!Event} e
+   * @private
+   */
+  handleInput_(e) {
+    const target = dev().assertElement(e.target);
+    if (!this.isDateField_(target) || target.type == 'hidden') {
+      return;
+    }
+
+    const property =
+        (target === this.dateField_ ? 'date' :
+          (target === this.startDateField_ ? 'startDate' :
+            (target === this.endDateField_ ? 'endDate' : '')));
+    const moment = this.createMoment_(target.value);
+    if (moment.isValid()) {
+      this.setState_({[property]: moment});
+    }
+  }
+
+  /**
+   * Handle changing values of the attached input fields, and the hotkey for
+   * closing the date picker.
+   * @param {!Event} e
+   */
+  handleKeydown_(e) {
+    const target = dev().assertElement(e.target);
+    if (this.isDateField_(target)) {
+      this.handleInputKeydown_(e);
+    } else {
+      this.handleDocumentKeydown_(e);
```

Comment:
`return this.` for both cases otherwise `return true` of `handleDocumentKeydown` won't do anything.